cpp/src/strings/strings_column_view.cu

```diff
@@ -57,113 +55,37 @@ size_type strings_column_view::chars_size() const noexcept
-  if (buffer_size == 0) {
-    std::cout << "all " << count << " strings are null\n";
-    return;
-  }
-  rmm::device_vector<char> buffer(buffer_size, 0);  // allocate and pre-null-terminate
-  char* d_buffer = buffer.data().get();
-  // copy strings into output buffer
-  thrust::for_each_n(
-    thrust::device,
-    thrust::make_counting_iterator<size_type>(0),
-    count,
-    [d_column, max_width, first, d_output_offsets, d_buffer] __device__(size_type idx) {
-      if (d_column.is_null(first + idx)) return;
-      string_view d_str = d_column.element<string_view>(first + idx);
-      size_type bytes   = d_str.size_bytes();
-      if ((max_width > 0) && (d_str.length() > max_width)) bytes = d_str.byte_offset(max_width);
-      memcpy(d_buffer + d_output_offsets[idx], d_str.data(), bytes);
-    });
-
-  // copy output buffer to host
-  std::vector<size_type> h_offsets(count + 1);
-  CUDA_TRY(cudaMemcpy(
-    h_offsets.data(), d_output_offsets, (count + 1) * sizeof(size_type), cudaMemcpyDeviceToHost));
-  std::vector<char> h_buffer(buffer_size);
-  CUDA_TRY(cudaMemcpy(h_buffer.data(), d_buffer, buffer_size, cudaMemcpyDeviceToHost));
 
-  // print out the strings to stdout
-  for (size_type idx = 0; idx < count; ++idx) {
-    size_type offset = h_offsets[idx];
-    size_type length = h_offsets[idx + 1] - offset;
-    std::cout << idx << ":";
-    if (length)
-      std::cout << "[" << std::string(h_buffer.data() + offset) << "]";
-    else
-      std::cout << "<null>";
-    std::cout << delimiter;
-  }
-}
-
-//
-std::pair<rmm::device_vector<char>, rmm::device_vector<size_type>> create_offsets(
+std::pair<rmm::device_uvector<char>, rmm::device_uvector<size_type>> create_offsets(
   strings_column_view const& strings,
   rmm::cuda_stream_view stream,
   rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
-  size_type count          = strings.size();
-  const int32_t* d_offsets = strings.offsets().data<int32_t>();
+  size_type const count = strings.size();
+
+  auto d_offsets = strings.offsets().data<int32_t>();
   d_offsets += strings.offset();  // nvbug-2808421 : do not combine with the previous line
-  int32_t first = 0;
-  CUDA_TRY(
-    cudaMemcpyAsync(&first, d_offsets, sizeof(int32_t), cudaMemcpyDeviceToHost, stream.value()));
-  rmm::device_vector<size_type> offsets(count + 1);
+
+  rmm::device_uvector<size_type> offsets(count + 1, stream);
   // normalize the offset values for the column offset
-  thrust::transform(
-    rmm::exec_policy(stream),
-    d_offsets,
-    d_offsets + count + 1,
-    offsets.begin(),
-    [first] __device__(int32_t offset) { return static_cast<size_type>(offset - first); });
+  thrust::transform(rmm::exec_policy(stream),
+                    d_offsets,
+                    d_offsets + count + 1,
+                    offsets.begin(),
+                    [d_offsets] __device__(int32_t offset) {
+                      return static_cast<size_type>(offset - d_offsets[0]);
+                    });
```

Comment:
Would this work, or does it result in dereferencing d_offsets on the host? (Not a big deal, just curious.)

```suggestion
                    [base = d_offsets[0]] __device__(int32_t offset) {
                      return static_cast<size_type>(offset - base);
                    });
```

Reply:
Yes, this will fail since it does the assignment in the host.